[Draft] Not That Trivial Resharding (basic infrastructure)

azat opened the pull request:
This is the draft of the re-sharding via ReplicatedMergeTree (as described in https://github.com/ClickHouse/ClickHouse/issues/45766).

### Plan

I've splitted implementation into smaller peaces:

- basic infrastructure
- pre-alpha version
- alpha version
- production

*I have checklist in my notes, that I can share later, once we will agree on the overall idea*

### Basic infrastructure

The main goal of this PR is to prepare some basic infrastructure with everything working in general (i.e. basic SELECT/INSERT, replication protocol, ALTERs, re-sharding) so that it can be merged as a first step and later the work can be done on top. This will allow to:
- make PR smaller, and hence review easier (at least I hope)
- reduce probability of conflicts
- involve more people to work on it later (there many things that can be done in parallel actually, once the infrastructure will be there)

Known tech debt:
- proper SELECTs support (i.e. now it does not work for secure connections, and if there is user/password)
I think that this bits can be done on top, not spent time on them.
- there are also some tests flakiness due to SYNC CLUSTER syncs only specific replica

### Implementation notes

I've tried to keep as less changes as possible in the common code (hence sometimes you may find "odd" code, or I avoided introducing some new SYSTEM queries in this PR, that's how `SYSTEM SYNC REPLICA CLUSTER` appeared), i.e. not modifying `StorageReplicatedMergeTree` a lot, but keep code in a separate classes.

**Changes should be 100% backward compatible if the cluster mode is not used.**

So this PR includes:
- basic SELECT support (but does not work when credential differs, or with secure protocol) - SELECT just gets the last cluster partitions map and send queries to those replicas with filter by specific partition
- basic INSERT support - split blocks on INSERT and send then to different replicas synchronously, when the block created first it is assigned to `cluster_replication_factor` replicas
- ALTERs/mutations/merges should work correctly (as far as I'm aware...)
- basic re-sharding support (it does not have any policies, right now the algorithm is pretty dumb, just distributed partitions until all replicas will have equal number of them)

#### Coordinator changes

It introduces concept of "cluster partition", it is stored in the `zookeeper_path/block_numbers/X`.
Also there is some meta information that is required, and it is stored under `zookeeper_path/cluster/`.
And now each replication entry has list of replicas that should execute this log entry (it is filtered only on the moving it form log to queue).

#### Settings

- table
  - `cluster`
  - `cluster_replicaton_factor`
- query
  - `cluster_query_shards`

#### Queries

- `SYSTEM SYNC REPLICA CLUSTER`
- `SYSTEM DROP CLUSTER REPLICA`

#### Introspection

- `system.cluster_partitions`

### Notes for reviewers

Since this is a draft, _please do not look thoroughly_, I'm sharing this to get some **feedback on the overall implementation and also coordinator communication**. I will share some of mine concerns (but maybe you will find something else):
- the main concern is that partition **migration is synchronous**, only after it done `DROP_RANGE` could be scheduled (and also now since DROP_RANGE is deferred you can get a race between removing the range of old parts during cloning partition and excecuting DROP_RANGE of the already migrated data from source replica, not a race, but assert that the state modified twice)
- right now I'm using some existing zookeeper nodes (i.e. `zookeeper_path/block_numbers/X` to store information about cluster partition), and I'm not sure that this is a good idea

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Not That Trivial Resharding

### Documentation entry for user-facing changes
TBD

#### CI Settings (Only check the boxes if you know what you are doing):
- [x] <!---woolen_wolfdog--> Woolen Wolfdog

Cc: @tavplubix

azat committed:
Fix simple count() query for ReplicatedMergeTree cluster mode

Before this patch it leads to this exception:

  Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column key in block. There are only columns: _partition_id: While executing Remote. (NOT_FOUND_COLUMN_IN_BLOCK)

The problem is that intiial query required {"key"} columns, while the
queries on replicas has {"_partition_id"}, and so it thinks that it does
not need any column, while actually it is a virtual column, so it still
needs another column, that will be added in ReadFromMergeTree

And anyway from the remote node it will got only "_partition_id" column
(due to lack of projections), but initiator expects "key".

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Initial cluster support for ReplicatedMergeTree infrastucture

This patch contains:
====================

- Distributed SELECT from the ReplicatedMergeTree with cluster=1
- Distributed INSERT from the ReplicatedMergeTree with cluster=1

Implementaiton details
======================

TBD

Query Settings
==============

- cluster_query_shards

MergeTree Settings
==================

- cluster=false
- cluster_replication_factor=1

ZooKeeper
=========

- /replicas/$/host -- information about node (disk space, labels) (*in future*)
- /block_numbers/$ -- now contains replicas list

Changelog:
- Fix ALTER for replicated cluster mode (ALTER_METADATA should be
  executed on all replicas)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Use storage specific virtual columns in TreeRewriterResult::collectUsedColumns()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
[RFC] Require at least one physical column in planner

This was an attempt to fix _partition_id filter for replicated-cluster,
but this is not enough still, since analyzer adds idnex to the columns
and so if the order will differs, then the header will not match, and
this is what happens now.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Initial re-sharding support

This is just a set of "C++ scripts" as the initial re-sharding
implementation.

Sadly but, as usual in big changes, it is pretty hard to avoid changing
something else, and this patch is not an exception from this... (some
logging, interfaces, tests).

Changelog:
- simple implementation with fetchPart() + DROP_RANGE
- deferred DROP_RANGE
- Distributor with FSM
- restore from coordinator
- clonePartition and friends
- cleaner API for ReplicatedMergeTreeClusterPartition
- REVERT_MIGRATE_PARTITION
- allow multiple partitions at once
- check version of the cluster/distributor node to avoid parallel re-sharding assignments
- proper wait of distributor on shutdown to avoid UAF
- tests with parallel replica restarts and fixes for it (uneven
  distribution after re-sharding)
- Remove ReplicatedMergeTreeClusterDistributor::migratePartition()
- Even distribution on INSERT
- Introduce ReplicatedMergeTreeClusterPartitionSelector
- ReplicatedMergeTreeClusterPartitionSelector::allocateParitition()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Split StorageReplicatedMergeTree::cloneReplicaIfNeeded() into helpers

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Split StorageReplicatedMergeTree::cloneReplica() into helpers

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Implement SYSTEM SYNC REPLICA db.table CLUSTER mode

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Initial cluster partition clone

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add default values for ReplicatedMergeTreeAddress

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Filter out cluster replicas for which we do not have info

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: add test with mutations

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Precise partitions_per_replica (fixes redundant CLONEs)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Allow removing replicas that is under migration

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Wait until partitions will be migrated from the dropped replica

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
SYSTEM DROP CLUSTER REPLICA

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add CLUSTER_SYNC replication command

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix ReplicatedMergeTreeClusterPartition::removeReplica()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Process cluster balancing more instantly

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not process empty parts

Likely should fix:

    Cannot process partition 3 (all_replicas: [1, 2, 4], active_replicas: [1, 2], MIGRATING 2 -> 4, version: 1), will revert: Code: 233. DB::Exception: Unexpected part name:

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Shutdown cluster in flushAndPrepareForShutdown()

So that we do not need to look at is_dropped/is_shutdown storage flags
inside balancer.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix reverting of partitions under migration/clone

Triggered by 03011_replicated_cluster_readonly_re_sharding

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: fix 03014_replicated_cluster_table_flaps (by adding one more SYNC REPLICA)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove one logical error that is possible for lost replicas

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: fix typo in 03015_replicated_cluster_mutations

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: add 03017_replicated_cluster_mutations_stress

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Ignore log_pointer for removed cluster replicas

Fixes: 03017_replicated_cluster_mutations_stress
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Check cluster partition version on mutation/merge entries creation and use active replicas

Otherwise it is possible to go into a sitatuion when mutation entry will
be created for the partition that will be removed during migration, and
never finished:

    │ 2023-09-29 17:43:54.778175 │      8939 │ default.data_r2 (Cluster)                                                        │ Loading partition from coordinator: 9 (all_replicas: [3, 4, 2], active_replicas: [3, 4], MIGRATING 3 -> 2, version: 7)
    │ 2023-09-29 17:43:54.778266 │      9111 │ default.data_r2 (ClusterBalancer)                                                │ Source replica does not have part 9_0_0_0_3. Removing it from ZooKeeper.
    │ 2023-09-29 17:43:54.778271 │      9111 │ default.data_r2 (ClusterBalancer)                                                │ Source replica does not have part 9_0_0_0_2. Removing it from ZooKeeper.
    │ 2023-09-29 17:43:54.778275 │      9111 │ default.data_r2 (ClusterBalancer)                                                │ Source replica does not have part 9_0_0_0_1. Removing it from ZooKeeper.
    │ 2023-09-29 17:43:54.778280 │      9111 │ default.data_r2 (ClusterBalancer)                                                │ Source replica does not have part 9_0_0_0_4. Removing it from ZooKeeper.
    │ 2023-09-29 17:43:54.778284 │      9111 │ default.data_r2 (ClusterBalancer)                                                │ Source replica does not have part 9_0_0_0_5. Removing it from ZooKeeper.
    │ 2023-09-29 17:43:54.778410 │      8939 │ default.data_r2 (ReplicatedMergeTreeQueue)                                       │ Will apply 20 mutations and mutate part 9_0_0_0_5 to version 25 (the last version is 25)
    │ 2023-09-29 17:43:54.778684 │      8939 │ default.data_r2 (d966de30-d7b7-404e-ba59-2b5e63072f44)                           │ Created log entry for mutation 9_0_0_0_25
    │ 2023-09-29 17:43:54.778694 │      9111 │ default.data_r2 (ClusterBalancer)                                                │ Source replica does not have part 9_0_0_0_5. Removing it from working set.
    │ 2023-09-29 17:43:54.778701 │      9111 │ default.data_r2 (ClusterBalancer)                                                │ Detaching 9_0_0_0_5

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not try to revert on session expired

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not retry reverts on session expired

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Allow fetchiing from dropped cluster replica

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Change severity for some logging

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Try to fix balancer stall

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: fix 03015_replicated_cluster_mutations

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Prune old partitions on SYNC REPLICA CLUSTER and adjust tests

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix readonly for non-cluster rmt

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add "Cluster sync took" log message

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not check log version on updating cluster partitions map

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Update cluster partition and clone partition in one tx

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove extra CV in StorageReplicatedMergeTree::cloneMetadataIfNeeded()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Pass single zookeeper instance

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove one FIXME, this should be OK

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Check log_pointer version correctly during cloning partition

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: ignore error "Cannot process partition ... will restart"

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Process log entries during cloning partition as well

This should avoid assigning different mutations, like:

    (CONNECTED [127.1:9181]) /> get /tables/test_yys1u8oj/data/log/log-0000000791
    format version: 8
    create_time: 2023-11-30 10:59:58
    source replica: 2
    block_id:
    log_entry_id:
    replicas
    2
    3

    mutate
    2_0_0_0_91
    to
    2_0_0_0_92

    (CONNECTED [127.1:9181]) /> get /tables/test_yys1u8oj/data/log/log-0000000898
    format version: 8
    create_time: 2023-11-30 11:00:01
    source replica: 1
    block_id:
    log_entry_id:
    replicas
    4
    1

    mutate
    2_0_0_0_91
    to
    2_0_0_0_102

And also copy proper entries so that it will know the history of the
part and fetch correct part, i.e. avoid situation like endless "Log
entry for part X or covering part is not pulled from log to
queue yet." for GET_PART:

   2023-11-30 11:00:01.094397      21172   test_yys1u8oj.data_r2 (c1c2713e-f48d-42ad-9327-f4ba7d17130f)    Will not fetch part 2_0_0_0_102 covering 2_0_0_0_91. Log entry for part 2_0_0_0_102 or covering part is not pulled from log to queue yet.
   2023-11-30 11:00:01.094617      21172   test_yys1u8oj.data_r2 (c1c2713e-f48d-42ad-9327-f4ba7d17130f)    Code: 234. DB::Exception: No active replica has part 2_0_0_0_91 or covering part (cannot execute queue-0000000710: GET_PART with virtual parts [2_0_0_0_91]). (NO_REPLICA_HAS_PART)

v2: remove REPLICA_STATUS_CHANGED - it is not correct to ignore log
entries in any way, since this creates parts without the history, and
breaks mutations and fetches
v3: Filter log entries by replicas

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Assign entries on all cluster replicas (including inactive)

Otherwise we will loose some changes during migration and this will
cause different mutations assigned:

    SELECT
        event_time,
        path,
        data
    FROM system.zookeeper_log
    WHERE (type = 'Response') AND (op_num = 'Create') AND (data LIKE '%mutate%') AND (path LIKE '%/log/%') AND (error = 'ZOK') AND (data LIKE '%mutate\n1_0_0_0_81\n%')
    ORDER BY event_time ASC
    LIMIT 10

    Query id: 2488f06a-f2e1-46bd-9acf-26f9a648ddcd

    ↘ Progress: 617.27 thousand rows, 65.94 MB (8.04 million rows/s., 858.87 MB/s.)  99%Row 1:
    ──────
    event_time: 2023-11-30 20:40:33.118551
    path:       /tables/test_g8yuxfij/data/log/log-
    data:       format version: 8
    create_time: 2023-11-30 20:40:33
    source replica: 2
    block_id:
    log_entry_id:
    replicas
    1
    2

    mutate
    1_0_0_0_81
    to
    1_0_0_0_82

    Row 2:
    ──────
    event_time: 2023-11-30 20:40:33.155124
    path:       /tables/test_g8yuxfij/data/log/log-
    data:       format version: 8
    create_time: 2023-11-30 20:40:33
    source replica: 4
    block_id:
    log_entry_id:
    replicas
    1
    4

    mutate
    1_0_0_0_81
    to
    1_0_0_0_83

    2 rows in set. Elapsed: 0.077 sec. Processed 617.27 thousand rows, 65.94 MB (8.02 million rows/s., 856.57 MB/s.)
    Peak memory usage: 1.42 MiB.

Log:

    2023-11-30 20:40:33.112819      23597   test_g8yuxfij.data_r4 (ClusterBalancer) Enqueued 1 fetches after mimic partition 1: 1_0_0_0_81
    2023-11-30 20:40:33.121317      23839   test_g8yuxfij.data_r1 (ReplicatedMergeTreeQueue)        Insert entry queue-0000000266 to queue with type MUTATE_PART with virtual parts [1_0_0_0_82] replicas [1, 2]

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Revert partition migration/clone only if the destionation is currently lost replica

Otherwise other replicas in cluster may cancel migrations from this
replica and this could happen over and over.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: add overlap mutations check in 03017_replicated_cluster_mutations_stress

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix intersecting mutations (do not assign mutations for not complete partitions)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove duplicated info form log message (info about replicas duplicated)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: Add long tag for 03017_replicated_cluster_mutations_stress

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add fixme for LOGICAL_ERROR due to trying to drop partition multiple times

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add some comments about removePartsFromWorkingSet() during clonePartition()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: add simple unittest for ActiveDataPartSet

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Simplify ReplicatedMergeTreeClusterBalancer::revert

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add more comments/fix wordings in ReplicatedMergeTreeClusterPartitionSelector

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not try to cleanup partitions that is under re-sharding

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Check cluster partition version during INSERT

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #58132 from azat/ClickHouse:replicated-cluster -->
*This is an automated comment for commit f780e5fc7a00e8972adcf4c6c49c27f7c9cf1f92 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58132/f780e5fc7a00e8972adcf4c6c49c27f7c9cf1f92/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58132/f780e5fc7a00e8972adcf4c6c49c27f7c9cf1f92/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

azat committed:
tests: fix shellcheck for 03011_replicated_cluster_readonly_re_sharding

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
docs: fix spellcheck

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add missing LOGICAL_ERROR into ReplicatedMergeTreeClusterPartitionSelector

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix typos in code comments

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not check for cluster+allow_remote_fs_zero_copy_replication (to unbreak CI)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Revert "[RFC] Require at least one physical column in planner" - to ubreak CI

This reverts commit 9916575b830929cc7db10c9b8ddfda0f0e090ad8.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Update description for mergetree setting "cluster" to fix 01221_system_settings

CI: https://s3.amazonaws.com/clickhouse-test-reports/58132/8a73f9e82c31a113ec5a56ef9a211722e94f3e41/fast_test.html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Revert "Use storage specific virtual columns in TreeRewriterResult::collectUsedColumns()"

This breaks some tests:
- 02707_clickhouse_local_implicit_file_table_function
- 02722_database_filesystem

CI: https://s3.amazonaws.com/clickhouse-test-reports/58132/8a73f9e82c31a113ec5a56ef9a211722e94f3e41/fast_test.html

This reverts commit 9314306296b229efb89035c42b091004493831ea.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Merge remote-tracking branch 'u/master' into replicated-cluster

* u/master: (5143 commits)

Conflicts:
	src/Interpreters/TreeRewriter.cpp
	src/Parsers/ASTSystemQuery.cpp
	src/Parsers/ParserSystemQuery.cpp
	src/Storages/StorageReplicatedMergeTree.cpp
	src/Storages/StorageReplicatedMergeTree.h
	src/Storages/System/attachSystemTables.cpp

Logical fixes:
- supportsTrivialCountOptimization() now has arguments
- around LoggerPtr

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
replicated-cluster: do not use indexHint() for filtering

It has been broken after merging with upstream

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Merge remote-tracking branch 'u/master' into replicated-cluster

azat committed:
Use PREWHERE for filtering by _partition_id instead of WHERE

WHERE is broken:

    root@833db487d7f6:~# chc --allow_experimental_analyzer=0 --optimize_trivial_count_query=0 --stage fetch_columns -q "select count() from data where _partition_id != 'all'" --format CSVWithNames --send_logs_level=test
    [833db487d7f6] 2024.04.16 12:12:14.734529 [ 19075 ] {283114d2-8b2c-4f6a-b4bb-85a29d787622} <Debug> default.data (b9dd6bfb-36af-4752-9348-a86282e3ad41) (SelectExecutor): Selected 1/1 parts by partition key, 1 parts by primary key, 1/1 marks by primary key, 1 marks to read from 1 ranges
    [833db487d7f6] 2024.04.16 12:12:14.735358 [ 19075 ] {283114d2-8b2c-4f6a-b4bb-85a29d787622} <Debug> executeQuery: Read 1 rows, 17.00 B in 0.001498 sec., 667.5567423230975 rows/sec., 11.08 KiB/sec.
    "number","_partition_id"
    0,"all"

    root@833db487d7f6:~# chc --allow_experimental_analyzer=0 --optimize_trivial_count_query=0 --stage fetch_columns -q "select count() from data prewhere _partition_id != 'all'" --format CSVWithNames --send_logs_level=test
    [833db487d7f6] 2024.04.16 12:13:15.376555 [ 19075 ] {dedf2197-a909-4257-a686-e9acdce96ec4} <Debug> default.data (b9dd6bfb-36af-4752-9348-a86282e3ad41) (SelectExecutor): Selected 0/0 parts by partition key, 0 parts by primary key, 0/0 marks by primary key, 0 marks to read from 0 ranges
    "number"

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Use overridden context for obtaining sample block

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Merge remote-tracking branch 'u/master' into replicated-cluster

Conflicts:
- src/Storages/StorageReplicatedMergeTree.cpp - movePartitionToTable()

azat committed:
Allow default cluster back

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not try to query shards for mutations (isStorageTouchedByMutations())

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix 03008_replicated_cluster__partition_id_filter (WHERE _partition_id is broken)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Explicitly disable analyzer for replicated-cluster tests

Onelinears:

    $ for i in tests/queries/0_stateless/*replicated_cluster*.sql; do { echo -e "SET allow_experimental_analyzer=0; -- FIXME: analyzer is not supported yet\n"; cat $i; } | sponge $i; done
    $ for i in tests/queries/0_stateless/*replicated_cluster*.sh; do { echo -e '#!/usr/bin/env bash\n\nCLICKHOUSE_CLIENT_OPT+="--allow_experimental_analyzer=0" # FIXME: analyzer is not supported yet\n'; grep -v '#!/usr/bin/env bash' $i; } | sponge $i; done

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Disable analyzer for 02813_tree_rewriter_virtual_columns

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add cluster_query_shards into SettingsChangesHistory

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Improve error message in ReplicatedMergeTreeClusterBalancer::replicatePartition()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add some comments for ReplicatedMergeTreeClusterBalancer

tarunannapareddy commented:
@azat @tavplubix what is the status of this feature? is the community currently interested in this way of introducing the re-sharding feature for clickhouse?

azat committed:
Merge remote-tracking branch 'u/master' into replicated-cluster

Conflicts:
    src/Core/SettingsChangesHistory.h
    src/Interpreters/ClusterProxy/executeQuery.cpp
    src/Interpreters/ClusterProxy/executeQuery.h
    src/Storages/StorageReplicatedMergeTree.cpp

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: add no-fasttest for replicated_cluster tests that could fail

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Merge remote-tracking branch 'u/master' into replicated-cluster

Conflicts:
    src/Core/SettingsChangesHistory.h

azat committed:
Merge remote-tracking branch 'u/master' into replicated-cluster

Ted-Jiang commented:
@azat Sorry for disturb you. May I ask what the progress in this Resharding implements? I want to contribute this feature together to make this go faster.

azat commented:
I would love to continue, and indeed, the idea is to provide some basic infrastructure for this, and after this, I could really use some help (there are lots of things would need to be done), but, there are couple of problems - I started this PR almost year ago and lots of things since then had been changed (i.e. I need to implement support for analyzer case) but one of the things that is not depends from me is the review, since I want to make sure that this is the best way to going forward.

I'm not saying that I'm going to abandon it, but I guess I need push more "energy" into this.

Ted-Jiang commented:
> I'm not saying that I'm going to abandon it, but I guess I need push more "energy" into this.
Thanks for your kindly quick reply 👍  , As i am newbee in clickhouse community, will try my best to involve in.   
@alexey-milovidov Could you give this some advice on this feature from core team's perspective? ❤️

alexey-milovidov commented:
I need this feature, as described here: https://github.com/ClickHouse/ClickHouse/issues/45766

Ted-Jiang commented:
As `cluster_replication_factor ` i have question about the how two set this.

If i have only **_one cluster each has 4 replica_**:
       create table like `cluster=1, cluster_replication_factor=2;` 

what the difference with   **_two cluster each has 2 replica_** 
          `cluster=1, cluster_replication_factor=2;` and `cluster=2, cluster_replication_factor=2;` 🤔

Ted-Jiang commented:
@azat I look through this pr carefully, nice design and clean code 👍  

As the  name == "cluster" setting on mergeTree is readonly. I think one big scenarios of this feature is: `split existing ReplicatedMergeTree `, Can i assume the existing one is create with `cluster = 1 , cluster_replication_factor=replication_number`?
